docs/monitoring.md

```diff
@@ -19,24 +19,43 @@ prometheus:
 
 ## Alerting
 
-If the scanner fails to push resources to Snyk's backend, the reconciliation
+We recommend that kubernetes-scanner monitor the scanner for failures, possibly
+with (presumably non-paging) alerts, in order to inform Snyk of potential issues
+with your installation. Please report persistent errors to [Snyk
+support](https://support.snyk.io/).
+
+When the scanner fails to push resources to Snyk's backend, the reconciliation
 will fail. The controller-runtime will ensure a continuous retry (with backoff).
 The metric `controller_runtime_reconcile_errors_total` is incremented for each
-failure. We recommend collecting these metrics and setting alerts on elevated
-reconcile error ratio. An example prometheus alerting query, that would return
-data when the error ratio exceeds 1%:
+failure. While we recommend collecting these metrics, we recommend setting
+alerts on the age of the oldest reconcilation failure that was not subsequently
+successfully retried, rather than on an elevated reconcile error ratio. An
+example prometheus query that will alert when a resource has not reconciled for
+1 hour:
 
 ```
-sum(rate(controller_runtime_reconcile_errors_total[1m])) / sum(rate(controller_runtime_reconcile_total[1m])) > 0.1
+kubernetes_scanner_backend_oldest_failure_age_seconds > 3600
 ```
 
-The example threshold of 10% may seem high, but occasional failures are
-expected, and the backoff-retry mechanism may often resolve them. As such it's
-recommended to set a sufficiently long `for` rule on any Prometheus alerts to
-balance signal with noise. We are actively working on a mechanism to allow
-persistent reconciliation failures to rise above the noise and allow alerting to
-be improved.
+If you do wish to alert on the error ratio, this example query will alert when
+the error ratio exceeds 5%:
+
+```
+sum(rate(controller_runtime_reconcile_errors_total[1m])) / sum(rate(controller_runtime_reconcile_total[1m])) > 0.05
+```
+
+The example threshold of 5% may seem high, but occasional failures are
```

Comment:
Seeing that our current failure ratio is below 1% (except for startup), I'd say we should decrease that even further.

(Assuming the usage of a sufficiently long `for` setting)

My target value for this would be max 1%. 

If we turn this around and write it in terms of an SLA, saying that (at least) 99% of requests should succeed isn't that much, is it? 😄

Reply:
For SLOs (as opposed to SLAs), I've seen anywhere between 0.05% and 2% as typical for web service (which this isn't) error ratio alerting thresholds.

The problem is, I don't think we can base anything resembling an SLO off this metric, because it doesn't distinguish between transient errors that are subsequently retried successfully within a subjectively acceptable eventual consistency window, and persistent failures. That's why we created the worst-case resource / oldest unreconciled failure metric.

It was only after being guided by that metric that we spotted our recently-solved WAF configuration problem that likely would have impacted many users - we couldn't tell whether the retry noise was business as usual, or evidence of persistent failures at first.

I suppose by tuning `for`, many users could indeed set a more ambitious error ratio threshold and still achieve good signal noise (with `for` config avoiding the likely surge in rate limit errors at startup as the scanner "settles").

But, I really don't think this matters much: this PR changes our recommended client alerting strategy to use that worse-case resource metric, which I think we agree has much sharper signal:noise than the reconcile error ratio? I wouldn't against a PR dropping this to 1%, but honestly at the moment I wouldn't recommend anyone set up such alerts at all. Charts, sure. But not alerts. Let's see what our early users experience, and go from there?